test/test_domain_migration.go

```diff
@@ -0,0 +1,101 @@
+	stream, err := client.StreamEvents(controller.StreamEventsOptions{
+		ObjectTypes: []ct.EventType{ct.EventTypeDomainMigration},
+	}, events)
+	t.Assert(err, c.IsNil)
+	defer stream.Close()
+
+	dm := &ct.DomainMigration{
+		Domain: newDomain,
+	}
+
+	prevRouterRelease, err := client.GetAppRelease("router")
+	t.Assert(err, c.IsNil)
+
+	err = client.PutDomain(dm)
+	t.Assert(err, c.IsNil)
+
+	// created
+	var event ct.DomainMigrationEvent
+	e := <-events
+	t.Assert(json.Unmarshal(e.Data, &event), c.IsNil)
+	t.Assert(event.Error, c.Equals, "")
+	t.Assert(event.DomainMigration, c.Not(c.IsNil))
+	t.Assert(event.DomainMigration.ID, c.Equals, dm.ID)
+	t.Assert(event.DomainMigration.OldDomain, c.Equals, dm.OldDomain)
+	t.Assert(event.DomainMigration.Domain, c.Equals, dm.Domain)
+	t.Assert(event.DomainMigration.TLSCert, c.IsNil)
+	t.Assert(event.DomainMigration.OldTLSCert, c.Not(c.IsNil))
+	t.Assert(event.DomainMigration.CreatedAt, c.Not(c.IsNil))
+	t.Assert(event.DomainMigration.CreatedAt.Equal(*dm.CreatedAt), c.Equals, true)
+	t.Assert(event.DomainMigration.FinishedAt, c.IsNil)
+
+	// complete
+	e = <-events
+	t.Assert(json.Unmarshal(e.Data, &event), c.IsNil)
+	t.Assert(event.Error, c.Equals, "")
+	t.Assert(event.DomainMigration, c.Not(c.IsNil))
+	t.Assert(event.DomainMigration.ID, c.Equals, dm.ID)
+	t.Assert(event.DomainMigration.OldDomain, c.Equals, dm.OldDomain)
+	t.Assert(event.DomainMigration.Domain, c.Equals, dm.Domain)
+	t.Assert(event.DomainMigration.TLSCert, c.Not(c.IsNil))
+	t.Assert(event.DomainMigration.OldTLSCert, c.Not(c.IsNil))
+	t.Assert(event.DomainMigration.CreatedAt, c.Not(c.IsNil))
+	t.Assert(event.DomainMigration.CreatedAt.Equal(*dm.CreatedAt), c.Equals, true)
+	t.Assert(event.DomainMigration.FinishedAt, c.Not(c.IsNil))
+
+	cert := event.DomainMigration.TLSCert
+
+	controllerRelease, err := client.GetAppRelease("controller")
+	t.Assert(err, c.IsNil)
+	t.Assert(controllerRelease.Env["DEFAULT_ROUTE_DOMAIN"], c.Equals, newDomain)
+	t.Assert(controllerRelease.Env["CA_CERT"], c.Equals, cert.CACert)
+
+	routerRelease, err := client.GetAppRelease("router")
+	t.Assert(err, c.IsNil)
+	t.Assert(routerRelease.Env["TLSCERT"], c.Equals, cert.Cert)
+	t.Assert(routerRelease.Env["TLSKEY"], c.Not(c.Equals), prevRouterRelease.Env["TLSKEY"])
+
+	dashboardRelease, err := client.GetAppRelease("dashboard")
+	t.Assert(err, c.IsNil)
+	t.Assert(dashboardRelease.Env["DEFAULT_ROUTE_DOMAIN"], c.Equals, newDomain)
+	t.Assert(dashboardRelease.Env["CONTROLLER_DOMAIN"], c.Equals, fmt.Sprintf("controller.%s", newDomain))
+	t.Assert(dashboardRelease.Env["URL"], c.Equals, fmt.Sprintf("dashboard.%s", newDomain))
+	t.Assert(dashboardRelease.Env["CA_CERT"], c.Equals, cert.CACert)
+
+	res, err := (&http.Client{}).Get(fmt.Sprintf("http://controller.%s/ping", newDomain))
+	t.Assert(err, c.IsNil)
+	t.Assert(res.StatusCode, c.Equals, 200, c.Commentf("failed to ping controller"))
+
+	res, err = (&http.Client{}).Get(fmt.Sprintf("http://dashboard.%s/ping", newDomain))
```

Comment:
We will need to add `dashboard.$CLUSTER_DOMAIN` to `/etc/hosts` in CI for this to work (i.e. [here](https://github.com/flynn/flynn/blob/v20151013.0/test/cluster/cluster.go#L557-L559)).